File: modules/org/openteacher/logic/safeHtmlChecker/safeHtmlChecker.js
/*
	Copyright 2013, Marten de Vries

	This file is part of OpenTeacher.

	OpenTeacher is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

	OpenTeacher is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with OpenTeacher.  If not, see <http://www.gnu.org/licenses/>.
*/

var isSafeHtml = (function () {
	var whitelist = [
		"address", "b", "big", "blockquote", "br", "center", "cite",
		"code", "dd", "dfn", "div", "dl", "dt", "em", "font", "h1",
		"h2", "h3", "h4", "h5", "h6", "hr", "i", "li", "nobr", "ol",
		"p", "pre", "s", "samp", "small", "span", "strong", "sub",
		"sup", "table", "tbody", "td", "tfoot", "th", "thead", "tr",
		"tt", "u", "ul", "var"
	];

	return function (html) {
		var tagRe = new RegExp("<\\s*([a-zA-Z]+)(?: [^>]*)?/?>");
		remainingHtml = html;
		while (true) {
			var match = remainingHtml.match(tagRe);
			if (match) {
				var tag = match[1].toLowerCase();
				if (whitelist.indexOf(tag) === -1) {
					return false; //unsafe html
				}
				//+1 so the current tag is not matched again.
				var pos = remainingHtml.search(tagRe) + 1;
				remainingHtml = remainingHtml.substr(pos);
			} else {
				return true; //safe html
			}
		}
	};
}());

File: modules/org/openteacher/logic/safeHtmlChecker/safeHtmlChecker.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-

#	Copyright 2013, Marten de Vries
#
#	This file is part of OpenTeacher.
#
#	OpenTeacher is free software: you can redistribute it and/or modify
#	it under the terms of the GNU General Public License as published by
#	the Free Software Foundation, either version 3 of the License, or
#	(at your option) any later version.
#
#	OpenTeacher is distributed in the hope that it will be useful,
#	but WITHOUT ANY WARRANTY; without even the implied warranty of
#	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#	GNU General Public License for more details.
#
#	You should have received a copy of the GNU General Public License
#	along with OpenTeacher.  If not, see <http://www.gnu.org/licenses/>.

class SafeHtmlCheckerModule(object):
	def __init__(self, moduleManager, *args, **kwargs):
		super(SafeHtmlCheckerModule, self).__init__(*args, **kwargs)
		self._mm = moduleManager

		self.type = "safeHtmlChecker"
		self.requires = (
			self._mm.mods(type="javaScriptEvaluator"),
		)
		self.javaScriptImplementation = True

	def isSafeHtml(self, html):
		"""Checks if it's safe to load an untrusted piece of ``html``
		   into a user's browser. Uses a tag whitelist.

		"""
		return self._js["isSafeHtml"](html)

	def enable(self):
		self._modules = next(iter(self._mm.mods(type="modules")))
		self._js = self._modules.default(type="javaScriptEvaluator").createEvaluator()
		with open(self._mm.resourcePath("safeHtmlChecker.js")) as f:
			self.code = f.read()
		self._js.eval(self.code)

		self.active = True

	def disable(self):
		self.active = False

		del self._modules
		del self._js
		del self.code

def init(moduleManager):
	return SafeHtmlCheckerModule(moduleManager)
